### DISCUSSION TITLE
How to deal with interpreters?

### DISCUSSION THREAD

**Discussion by martin-braun** [objective-see/LuLu#606] [Q&A]
I stopped using Little Snitch and want to use Lulu. The experience is great, so thanks for the good product.

One thing that I miss though are rules that are defined based on the host who spawned the process in question.

For instance. An application, written in Java, will spawn the Java binary in the rule setup of Lulu. If two Applications are Java applications, their rules have to be shared, since those are Java rules, not specific Application rules.

The same limitation is present with other interpreters, such as Deno, Python, node, etc. 

Little Snitch allowed rules to be different based on the caller. So much so, that I could even influence `curl` rules for each script individually.

Can something like this be done in Lulu already? Or is it planned?

EDIT: I just see that the `java` rule is in fact for a binary embedded into the Application, so this might not be a problem. Still, I think this would be such a great addition.

**Comment by martin-braun** [comment]
Here is a real life example:

The Android SDK / Android Emulator will access the `java` version I setup properly via `sdkman`. This is the `java` in `$JAVA_HOME`:

<img width="871" alt="image" src="https://github.com/objective-see/LuLu/assets/24259245/94bfc0c8-302c-4cfd-af5c-bf923acb7668">

I have no power to say that it should be a rule only for the emulator / Android Studio / Android SDK, instead I can only whitelist the entire `java` binary making it open for other apps and scripts to exploit.

**Comment by objective-see** [comment]
Yes this would be good to have! 

I think the "easiest" option would be to have a rule option (for interpreters) that would allow you to create a rule for the interpreter + script (some what how now you can create a rule that is for process + endpoint). 

So you could make a rule for: `Python` + `<some script>`
...but that wouldn't match for `Python` + `<some other script>`

Thoughts?

**Reply by martin-braun** [reply]
A rule could just have an ancestor program path in additional to the regular program path. The program path would point to `/usr/local/Cellar/.../python` while the ancestor program path could point to `/Users/myuser/.local/bin/my-script.sh` or any other host application.

Little Snitch was able to determine the ancestor at runtime to even attach this information during automatic rule creation, something I would appreciate a lot.

Yes, the rule would combine both paths. Rules without an ancestor path would work on all calls of the binary.

**Reply by Pixel-Master** [reply]
Wouldn't the way Block-Block deals with this, also work for LuLu? 
<img width="857" alt="Bildschirmfoto 2025-01-27 um 20 01 58" src="https://github.com/user-attachments/assets/7356b113-c2dc-412b-9e5d-e6bb43a3fb24" />

Maybe a drop-down with every possible combination is a bit convoluted, a check-able Drop-down might work better. Like this
<img width="227" alt="SCR-20250127-rlzp" src="https://github.com/user-attachments/assets/b52075a0-a076-42a2-98c0-c789345ce568" />

**Reply by martin-braun** [reply]
Maybe we are overthinking it? Sure the way Little Snitch does it goes for maximum control, but the key issue is that I don't want to specify rules for Python or curl, but for the ancestor program.

If I launch some `telemetry.sh`, I want to block access to the internet on anything that `telemetry.sh` uses to do so, no matter if its `curl` or `wget`. If it uses both, LuLu shouldn't ask me twice. If I want to block only certain URLs I can always block access to the specific destination, not the process.

However, triggering only on root processes is hard, because everything is a subprocess of another process. I have no idea about the Apple API, but setting up rules for interpreters like `Python` or CLI tools like `curl` (which are used in scripts) feels so wrong.

On top of that, I get the dialogs over and over again for my scripts, because I updated the interpreter, yet the ancestor script was not updated at all. This is disturbing and makes working with LuLu rather a chore.

**Reply by objective-see** [reply]
Well in this case `telemetry.sh` isn't really the parent (ancestor program) ...it'd actually be the interpreter such as `bash` (or `zsh`). And given this fact, it's actually not trivial to determine what is the script as the real parent (e.g. `bash`, the interpreter) could have other arguments, both before and after the script. (Also the script could be specified relative to CWD). 

Moreover, what if a child (e.g `curl` or `wget`) itself spawns another child. In theory you'd likely want to block that too. Meaning now we have to take into the account the entire process hierarchy. 

The point being, this is all very nuanced. 

I'm leaning toward just adding extra logic for interpreters (or any process?) where rules can be created based on their arguments?
Example: `bash` is allowed if the argument is `telemetry.sh`

Or, maybe rules, yes based on the parent:
Example: `curl` is allowed if the parent is `bash` ...though this seems way to broad (as lot of malware may spawn `curl` via the shell.
